File: scripts/viz_lib.py
"""Hand detection + 21-keypoint visualization helpers.

Pure drawing; no model/IO logic. Consumes results already mapped back to
original-image pixel coordinates (i.e. post `unletterbox_bboxes` /
`affine_kpts`). Reused by bench_e2e.py for single-backend dumps and for
left|right hstack in `--backend both`.
"""
from __future__ import annotations
import numpy as np
import cv2


# MediaPipe 21-keypoint hand topology (wrist=0, thumb 1-4, index 5-8,
# middle 9-12, ring 13-16, pinky 17-20). 20 bones total.
HAND_SKELETON: tuple[tuple[int, int], ...] = (
    (0, 1), (1, 2), (2, 3), (3, 4),
    (0, 5), (5, 6), (6, 7), (7, 8),
    (5, 9), (9, 10), (10, 11), (11, 12),
    (9, 13), (13, 14), (14, 15), (15, 16),
    (13, 17), (0, 17), (17, 18), (18, 19), (19, 20),
)

# BGR — one color per finger; wrist shares thumb color.
_THUMB  = (0, 165, 255)   # orange
_INDEX  = (0, 255, 255)   # yellow
_MIDDLE = (0, 255, 0)     # green
_RING   = (255, 128, 0)   # cyan-ish blue
_PINKY  = (255, 0, 255)   # magenta

# index -> color, length 21
KPT_COLORS: tuple[tuple[int, int, int], ...] = (
    _THUMB,                                  # 0 wrist
    _THUMB,  _THUMB,  _THUMB,  _THUMB,       # 1-4
    _INDEX,  _INDEX,  _INDEX,  _INDEX,       # 5-8
    _MIDDLE, _MIDDLE, _MIDDLE, _MIDDLE,      # 9-12
    _RING,   _RING,   _RING,   _RING,        # 13-16
    _PINKY,  _PINKY,  _PINKY,  _PINKY,       # 17-20
)

BBOX_COLOR = (0, 255, 0)   # green


def draw_hands(
    img_bgr: np.ndarray,
    boxes: np.ndarray,
    kpts_per_pred: dict,
    *,
    scores: np.ndarray | None = None,
    label: str | None = None,
) -> np.ndarray:
    """Return a BGR copy of `img_bgr` with bboxes + 21 keypoints + skeleton.

    Args:
        img_bgr        : (H,W,3) uint8 BGR original-image frame.
        boxes          : (N,4) float xyxy in original image coords.
        kpts_per_pred  : {pi: (21,2) float} kpts in original image coords.
        scores         : (N,) optional detection confidence. If given, drawn
                         next to each bbox as 'det=0.87'.
        label          : optional top-left tag, e.g. 'ONNX' / 'RKNN'. Drawn
                         on a black filled rect for legibility under hstack.

    Always returns a NEW array (np.copy) so callers can freely hstack the
    result without aliasing back into the shared rec['img'].
    """
    vis = img_bgr.copy()

    for pi, box in enumerate(boxes):
        x1, y1, x2, y2 = (int(round(v)) for v in box[:4])
        cv2.rectangle(vis, (x1, y1), (x2, y2), BBOX_COLOR, 2)
        if scores is not None and pi < len(scores):
            txt = f"det={float(scores[pi]):.2f}"
            cv2.putText(vis, txt, (x1, max(y1 - 4, 12)),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.5, BBOX_COLOR, 1,
                        cv2.LINE_AA)

        kp = kpts_per_pred.get(pi)
        if kp is None:
            continue
        # skeleton first so dots overlay the lines cleanly
        for i, j in HAND_SKELETON:
            if i >= len(kp) or j >= len(kp):
                continue
            p1 = (int(round(kp[i, 0])), int(round(kp[i, 1])))
            p2 = (int(round(kp[j, 0])), int(round(kp[j, 1])))
            cv2.line(vis, p1, p2, KPT_COLORS[j], 1, cv2.LINE_AA)
        for k in range(min(21, len(kp))):
            cx, cy = int(round(kp[k, 0])), int(round(kp[k, 1]))
            cv2.circle(vis, (cx, cy), 3, KPT_COLORS[k], -1, cv2.LINE_AA)

    if label:
        # black filled rect + white text — survives white backgrounds
        font, scale_, thick = cv2.FONT_HERSHEY_SIMPLEX, 0.7, 2
        (tw, th), base = cv2.getTextSize(label, font, scale_, thick)
        pad = 4
        cv2.rectangle(vis, (0, 0), (tw + 2 * pad, th + 2 * pad + base),
                      (0, 0, 0), -1)
        cv2.putText(vis, label, (pad, th + pad), font, scale_,
                    (255, 255, 255), thick, cv2.LINE_AA)

    return vis

File: src/infer/tests/viz_lib.cpp
// 板端测试可视化实现; 字面对齐 scripts/viz_lib.py:draw_hands。
#include "viz_lib.hpp"

#include <opencv2/imgproc.hpp>

#include <algorithm>
#include <cmath>
#include <cstdio>

namespace hand_pipeline::viz {

namespace {

// Per-finger BGR. cv::Scalar 构造不是 constexpr, 用 const 即可。
const cv::Scalar kThumb (0,   165, 255);   // orange
const cv::Scalar kIndex (0,   255, 255);   // yellow
const cv::Scalar kMiddle(0,   255, 0);     // green
const cv::Scalar kRing  (255, 128, 0);     // cyan-ish blue
const cv::Scalar kPinky (255, 0,   255);   // magenta

inline int iround(float v) {
    return static_cast<int>(std::lround(v));
}

}  // namespace

const std::array<std::pair<int, int>, 21> kHandSkeleton = {{
    {0, 1},  {1, 2},   {2, 3},   {3, 4},
    {0, 5},  {5, 6},   {6, 7},   {7, 8},
    {5, 9},  {9, 10},  {10, 11}, {11, 12},
    {9, 13}, {13, 14}, {14, 15}, {15, 16},
    {13, 17}, {0, 17}, {17, 18}, {18, 19}, {19, 20},
}};

const std::array<cv::Scalar, kNumKpts> kKptColors = {{
    kThumb,                                    // 0 wrist
    kThumb,  kThumb,  kThumb,  kThumb,         // 1-4
    kIndex,  kIndex,  kIndex,  kIndex,         // 5-8
    kMiddle, kMiddle, kMiddle, kMiddle,        // 9-12
    kRing,   kRing,   kRing,   kRing,          // 13-16
    kPinky,  kPinky,  kPinky,  kPinky,         // 17-20
}};

const cv::Scalar kBboxColor(0, 255, 0);

cv::Mat DrawHands(
    const cv::Mat& img_bgr,
    const std::vector<Detection>& boxes,
    const std::vector<std::array<Keypoint, kNumKpts>>& kpts_per_pred,
    const std::string& label) {

    // Clone, mirrors Python np.copy(); 调用方共享 LoadedImage::img 安全。
    cv::Mat vis = img_bgr.clone();

    for (std::size_t pi = 0; pi < boxes.size(); ++pi) {
        const Detection& b = boxes[pi];
        const cv::Point p1(iround(b.x1), iround(b.y1));
        const cv::Point p2(iround(b.x2), iround(b.y2));
        cv::rectangle(vis, p1, p2, kBboxColor, 2);

        char buf[32];
        std::snprintf(buf, sizeof(buf), "det=%.2f", b.score);
        const cv::Point txt_org(p1.x, std::max(p1.y - 4, 12));
        cv::putText(vis, buf, txt_org,
                    cv::FONT_HERSHEY_SIMPLEX, 0.5, kBboxColor, 1,
                    cv::LINE_AA);

        if (pi >= kpts_per_pred.size()) continue;
        const auto& kp = kpts_per_pred[pi];

        // 骨架先画, 圆点后画 — 与 Python 顺序一致, 点压在线上更清晰。
        for (const auto& [i, j] : kHandSkeleton) {
            if (i >= kNumKpts || j >= kNumKpts) continue;
            const cv::Point a(iround(kp[i].x), iround(kp[i].y));
            const cv::Point b2(iround(kp[j].x), iround(kp[j].y));
            cv::line(vis, a, b2, kKptColors[j], 1, cv::LINE_AA);
        }
        for (int k = 0; k < kNumKpts; ++k) {
            cv::circle(vis, cv::Point(iround(kp[k].x), iround(kp[k].y)),
                       3, kKptColors[k], -1, cv::LINE_AA);
        }
    }

    if (!label.empty()) {
        const int font = cv::FONT_HERSHEY_SIMPLEX;
        const double scale_ = 0.7;
        const int thick = 2;
        int baseline = 0;
        const cv::Size ts = cv::getTextSize(label, font, scale_, thick,
                                            &baseline);
        const int pad = 4;
        cv::rectangle(vis,
                      cv::Point(0, 0),
                      cv::Point(ts.width + 2 * pad,
                                ts.height + 2 * pad + baseline),
                      cv::Scalar(0, 0, 0), cv::FILLED);
        cv::putText(vis, label, cv::Point(pad, ts.height + pad),
                    font, scale_, cv::Scalar(255, 255, 255), thick,
                    cv::LINE_AA);
    }

    return vis;
}

}  // namespace hand_pipeline::viz
